projects/concurrency/cache.go

```diff
@@ -0,0 +1,191 @@
+	list                 *LinkedList[K]
+}
+
+type LinkedList[K comparable] struct {
+	head *Node[K]
+	tail *Node[K]
+}
+
+func NewCache[K comparable, V any](entryLimit int) Cache[K, V] {
+	return Cache[K, V]{entryLimit: entryLimit, cache: make(map[K]cacheEntry[K, V]), list: &LinkedList[K]{}}
+}
+
+func (c *Cache[K, V]) Put(key K, value V) bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.totalWrites += 1
+
+	entry := cacheEntry[K, V]{value: value, timeStamp: time.Now()}
+
+	if value, ok := c.cache[key]; ok {
+		// means that it is already in the list, move to head
+		entry.numberOfTimesRead = value.numberOfTimesRead
+		c.cache[key] = entry
+		return ok
+	}
+
+	//inserting new item
+	if len(c.cache) == c.entryLimit {
+		// get key at the end of the list, delete from map and linked list
+		lastUsed := getLastRecentlyUsedEntry(c.cache)
+		if c.cache[lastUsed].numberOfTimesRead == 0 {
+			c.unReadEvictedEntries += 1
+		}
+		delete(c.cache, lastUsed)
+	}
+
+	// insert new item at the beginning of the list
+	c.cache[key] = entry
+	return false
+}
+
+func (c *Cache[K, V]) Get(key K) (*V, bool) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.totalReads += 1
+	entry, ok := c.cache[key]
+	if !ok {
+		return nil, ok
+	}
+	entry.timeStamp = time.Now()
+	entry.numberOfTimesRead++
+	c.cache[key] = entry
+	c.successfulReads += 1
+	// retrieve node from map and move to head
+	return &entry.value, ok
+}
+
+func (c *Cache[K, V]) GetStatistics() Statistics {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	var hitRate float64
+	if c.totalReads != 0 {
+		result := float64(c.successfulReads) / float64(c.totalReads) * 100
+		hitRate = roundFloat(result, 2)
+	}
+	//calculate average reads
+	totalReadsInCache := 0
+	for _, value := range c.cache {
+		totalReadsInCache += value.numberOfTimesRead
+	}
+	averageReads := float64(totalReadsInCache) / float64(len(c.cache))
+
+	return Statistics{
+		hitRate:             hitRate,
+		averageReads:        roundFloat(averageReads, 2),
+		totalReadsAndWrites: c.totalReads + c.totalWrites,
+		unReadEntries:       c.unReadEvictedEntries + unReadEntriesInMap(c.cache),
+	}
+}
+
+func getLastRecentlyUsedEntry[K comparable, V any](entries map[K]cacheEntry[K, V]) K {
+	var lastUsedKey K
+	var readTime = time.Date(9999, 1, 1, 0, 0, 0, 0, time.UTC)
+
+	for key, value := range entries {
+		if value.timeStamp.Before(readTime) {
+			readTime = value.timeStamp
+			lastUsedKey = key
+		}
+	}
+	return lastUsedKey
+}
+
+func unReadEntriesInMap[K comparable, V any](entries map[K]cacheEntry[K, V]) int {
+	count := 0
+
+	for _, value := range entries {
+		if value.numberOfTimesRead == 0 {
+			count++
+		}
+	}
+	return count
+}
+
+func roundFloat(num float64, precision int) float64 {
+	multiplier := math.Pow(10.0, float64(precision))
+	return math.Round(num*multiplier) / multiplier
+}
+
+func (list *LinkedList[K]) moveToHead(node *Node[K]) {
+	if (node == list.head) {
+		return
+	}
+	if (list.head == nil) {
+		list.insertAtHead(node)
+		return
+	}
```

Comment:
These early returns make it really easy to know what these cases are for, and ignore them when reading the rest of the function - nice!